File: src/main/java/org/jcodec/common/model/TapeTimecode.java
package org.jcodec.common.model;

import static java.lang.String.format;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Tape timecode
 * 
 * @author The JCodec project
 * 
 */
public class TapeTimecode {
    public static final TapeTimecode ZERO_TAPE_TIMECODE = new TapeTimecode((short) 0, (byte) 0, (byte) 0, (byte) 0, false, 0);

    private final short hour;
    private final byte minute;
    private final byte second;
    private final byte frame;
    private final boolean dropFrame;
    private final int tapeFps;

    public TapeTimecode(short hour, byte minute, byte second, byte frame, boolean dropFrame, int tapeFps) {
        this.hour = hour;
        this.minute = minute;
        this.second = second;
        this.frame = frame;
        this.dropFrame = dropFrame;
        this.tapeFps = tapeFps;
    }

    public short getHour() {
        return hour;
    }

    public byte getMinute() {
        return minute;
    }

    public byte getSecond() {
        return second;
    }

    public byte getFrame() {
        return frame;
    }

    public boolean isDropFrame() {
        return dropFrame;
    }

    public int getTapeFps() {
        return tapeFps;
    }

    @Override
    public String toString() {
        String tcfmt = dropFrame ? "%02d:%02d:%02d;%02d" : "%02d:%02d:%02d:%02d";
        return format(tcfmt, hour, minute, second, frame);
    }

    public static TapeTimecode tapeTimecode(long frame, boolean dropFrame, int tapeFps) {
        if (dropFrame) {
            long D = frame / 17982;
            long M = frame % 17982;
            frame += 18 * D + 2 * ((M - 2) / 1798);
        }
        long sec = frame / tapeFps;
        return new TapeTimecode((short) (sec / 3600), (byte) ((sec / 60) % 60), (byte) (sec % 60),
                (byte) (frame % tapeFps), dropFrame, tapeFps);
    }
}

File: src/main/java/org/jcodec/common/model/TapeTimecode.kt
package org.jcodec.common.model

import org.jcodec.common.StringUtils

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Tape timecode
 *
 * @author The JCodec project
 */
class TapeTimecode(val hour: Short, val minute: Byte, val second: Byte, val frame: Byte, val isDropFrame: Boolean, val tapeFps: Int) {

    override fun toString(): String {
        return StringUtils.zeroPad2(hour.toInt()) + ":" +
                StringUtils.zeroPad2(minute.toInt()) + ":" +
                StringUtils.zeroPad2(second.toInt()) + (if (isDropFrame) ";" else ":") +
                StringUtils.zeroPad2(frame.toInt())
    }

    companion object {
        @JvmField
        val ZERO_TAPE_TIMECODE = TapeTimecode(0.toShort(), 0.toByte(), 0.toByte(), 0.toByte(), false, 0)
        @JvmStatic
        fun tapeTimecode(frame: Long, dropFrame: Boolean, tapeFps: Int): TapeTimecode {
            var frame = frame
            if (dropFrame) {
                val D = frame / 17982
                val M = frame % 17982
                frame += 18 * D + 2 * ((M - 2) / 1798)
            }
            val sec = frame / tapeFps
            return TapeTimecode((sec / 3600).toShort(), (sec / 60 % 60).toByte(), (sec % 60).toByte(),
                    (frame % tapeFps).toByte(), dropFrame, tapeFps)
        }
    }

}